Unity

emilnorsker opened the pull request:
## Description
This PR introduces unity as one of our build targets. 
There are a lot of issues this far especially with our DEV and CI environment. 
The majority of this MR is focused on fixing that.


## TYPE
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update


Nix develop now creates a unity executable, but due to licensing and the sandboxing in nix, it is not currently possible to activate and return a license in the environment. Also for personal (I have not tested on professional) licenses, they are tied to a specific machine id, which is different from a sandboxed environemnt when compared to the original users' env. This problem will problably be exacerbated by running this on ephemeral github runners.   

This is currently stopping us from running our standard pipeline in CI, which basically means that we cant ensure the stability of the plugin right now. 

In spite of these limitations, there is a decent DEV environment running. To access this you will need a unity account and you will need to log in at least once and add a license to your computer. but after that it should not be required anymore.

Testing locally can be done by `cd nobody/unity` -> `nix develop` -> `just test`.
The same goes for running a unity with the package built in. Simply `cd nobody/unity` -> `nix develop` -> `just run-unity` will start the editor inside a temporary disposable unity project. 

The model can then be added as a component:

![250306_12h16m17s_screenshot](https://github.com/user-attachments/assets/80c0efb6-2d58-4be3-81d4-a5250aefbc76)

emilnorsker commented:
Results from tests:
```
📊 Test Results Summary:
====================
✨ Total Tests: 4
✅ Passed: 4
❌ Failed: 0
⏱️  Duration: 4.320658 seconds


[norsker@nixos:~/Documents/nobodywho/nobodywho/unity]$ git status
On branch unity
Your branch is up to date with 'origin/unity'.

nothing to commit, working tree clean
```

AsbjornOlling commented:
This is looking good! Here are some first-glance notes:

You're probably aware, but it seems like there are a few things not implemented:

- context size. seems like it's hardcoded to 4096
- sampler configuration
- stop words
- embeddings

I found a few potential errors:
- what happens if the error returned is longer than the StringBuilder size (256)? I think several of our errors are longer than 256 bytes. we should make the error string buffer way longer, just to fit our own error messages. this also seems like it could be a overflow bug. we should also do a bounds-check on the rust side, and a short-enough fallback error message, just to prevent the overflow.
- `destroy_chat_worker` is declared on the C#-side, but not on the rust side. It seems like the model and context are always leaked, with no way of cleaning it up. Maybe we could do it w/ Unity's `OnDestroy`?
- the return value of `send_prompt` seems to be ignored. it returns `bool` in rust, but it's declared as returning `void` on the C# side.
- why use `Box` in `get_model`? `llm::Model` is already an `Arc<LlamaModel>`. Could it be sufficient to use `Arc::into_raw` and `Arc::from_raw`, rather than boxing and unboxing the `Arc<LlamaModel>`? There could be an issue related to maintaining refcounts.. What we do on the godot side is to pass in a reference to an struct which implements the `get_model` method, maybe that would let us maintain refcounts?

Some potential improvements:
- does unity provide a file picker for properties on nodes that are files? Maybe we could replace the type of `modelPath` to be some kind of File?
- it looks like we currently expect the model file path to be relative to `streamingAssetsPath`. On the godot side, it is possible to specify a path in `user://`, `res://`, or an absolute path on the filesystem. Would it be possible to do something more general like that? It also [looks like](https://docs.unity3d.com/6000.0/Documentation/Manual/StreamingAssets.html) streamingAssets paths might not work for android builds, since they seem to resolve to a `jar://` path, which I expect llama.cpp won't be able to work with.

Some stuff I found really cool:
- it looks like a really clean, small codebase. Not as much boilerplate as i had feared!
- looks like you got two-way FFI working! very cool that you can pass in functions from unity, to be called from rust. This might be very handy for generic tool-calling. Maybe it could also let us avoid doing anything in `Update`.

emilnorsker commented:
So after you comment, I researched what they do in the LLMUnity, in their [chatbot sample](https://github.com/undreamai/LLMUnity/tree/main/Samples~/ChatBot). The scene is structured like this: 
```
Scene: MainScene
├── LLM (GameObject)
│   └── NobodyWho.Model (Script)
│       - port: 13333
│       - contextSize: 8192
│       - temperature: 0.2
│       - topK: 40
│       - topP: 0.9
│       - stream: true
│
├── LLMCharacter (GameObject)
│   └── NobodyWho.Chat (Script)
│       - llm: {reference to LLM GameObject's Model component}
│       - host: localhost
│       - port: 13333
│       - temperature: 0.2
│       - stream: true
│       - playerName: "user"
│       - AIName: "assistant"
│       - prompt: "A chat between a curious human and an artificial intelligence assistant..."
│
└── ChatBot (GameObject)
    └── NobodyWho.ChatBot (Script)
        - chatContainer: {reference to ChatPanel}
        - llmCharacter: {reference to LLMCharacter's Chat component}
```

This approach is slightly different in the way they create a gameobject for each script. translated to Godot this is like creating a node with a script attached instead of loading a script or referencing a script. 
Our architecture works the same as you can just add the scripts to their separate gameobjects and connect them.

emilnorsker commented:
>    destroy_chat_worker is declared on the C#-side, but not on the rust side. It seems like the model and context are always leaked, with no way of cleaning it up. Maybe we could do it w/ Unity's OnDestroy?

OnDestroy was a great suggestion. this is implemented now.


>    the return value of send_prompt seems to be ignored. it returns bool in rust, but it's declared as returning void on the C# side.

Good catch, it was a leftover from one of the earlier iterations of error handling.

>    why use Box in get_model? llm::Model is already an Arc<LlamaModel>. Could it be sufficient to use Arc::into_raw and Arc::from_raw, rather than boxing and unboxing the Arc<LlamaModel>? There could be an issue related to maintaining refcounts.. What we do on the godot side is to pass in a reference to an struct which implements the get_model method, maybe that would let us maintain refcounts?

Yup, you are correct. I did not now how Arc works and I have a better understanding now. If you have the time i would appreciate an extra look on the new implementation. I think that it works correctly now - but i am a bit unsure if i got all the nuances of rusts memory management correct. 
The docs says:

> pub fn [into_raw](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.into_raw)(this: [Arc](https://doc.rust-lang.org/std/sync/struct.Arc.html)<T, A>) -> [*const T](https://doc.rust-lang.org/std/primitive.pointer.html)
Consumes the Arc, returning the wrapped pointer.
To avoid a memory leak the pointer must be converted back to an Arc using [Arc::from_raw](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.from_raw).

The NobodyWho.Model instance should keep a reference to the pointer as soon as we create a worker context. 
```csharp
    public class Model : MonoBehaviour {
        public string modelPath = "model.gguf";
        public bool useGpuIfAvailable = true;
        private IntPtr modelHandle;
        ...
```
`_workerContext = NativeBindings.create_chat_worker(model.GetModel(), systemPrompt);`

```
# in GetModel
var result = NativeBindings.get_model(fullPath, useGpuIfAvailable, errorBuffer);
...
modelHandle = result;
return modelHandle;
```

When we destroy the object we will then use the `Arc::from_raw` to get the last reference and let it go out of scope to be garbage collected:

```csharp
# Nobodywho.Model
    void OnDestroy() {
            NativeBindings.destroy_model(modelHandle);
        }
```
```rust
/// lib.rs
#[no_mangle]
pub extern "C" fn destroy_model(model: *mut c_void) {
    unsafe {
        let _: Arc<llm::Model> = Arc::from_raw(model as *const llm::Model); 
    }
}
```

emilnorsker commented:
proof of test. at HEAD: 4f19620
![image](https://github.com/user-attachments/assets/263c1998-1ebf-42c5-80e9-becd64e79502)

emilnorsker commented:
At least unit tests exists for the ffi code for unity. 
The integration test will stil have to be run by hand.. for now

```
nobodywho> running 1 test
nobodywho> test tests::test_create_chat_worker ... thread 'tests::test_create_chat_worker' panicked at unity/src/lib.rs:210:9:
nobodywho> assertion `left == right` failed: Model should be loaded successfully
nobodywho>   left: [77, 111, 100, 101, 108, 32, 110, 111, 116, 32, 102, 111, 117, 110, 100, 58, 32, 113, 119, 101, 110, 50, 46, 53, 45, 49, 46, 53, 98, 45, 105, 110, 115, 116, 114, 117, 99, 116, 45, 113, 52, 95, 48, 46, 103, 103, 117, 102]
nobodywho>  right: []
nobodywho> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
nobodywho> FAILED
nobodywho> 
nobodywho> failures:
nobodywho> 
nobodywho> failures:
nobodywho>     tests::test_create_chat_worker
nobodywho> 
nobodywho> test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
nobodywho> 
nobodywho> error: test failed, to rerun pass `-p unity-plugin --lib`
```

emilnorsker commented:
The following features are now implemented in unity:


# Chat completion 
Chat completion with history and context shifting is now implemented as in Godot.

- token streaming
- gbnf grammars
- context length
- stop words 
- function callback that works like signals
- responses are await-able with this interface:
```
        [Test]
        public async Task WhenInvokingSayWithSingleStopWord_ShouldStopAtStopWord() {
            string response = null;
            chat.stopWords = "fly";
            chat.ResetContext();

            response = await chat.Say("List these animals in alphabetical order: cat, dog, fly, lion, mouse");
}
```


NOTE: We have decided not to implement specialized sampler configurations. this is due to the low usage of the feature in Godot and the complexity of implementing it in the ffi interface.

# Embeddings

- embed text
- cosine similarity
- both callbacks and awaitable interface as with the chat completion.



# Other

- The code is quite heavily tested but only the cargo tests can be run in CI.
- There is another interface for starting the model from a node. Currently a contextworker is always created on the unity start function--`Start` is equal to Godot's `_ready` function-- meaning that it might be heavy to instantiate several Chats at once.

emilnorsker commented:
![image](https://github.com/user-attachments/assets/3ea18e3f-7a07-4431-a15d-d6d2e7d8d6f9)


![image](https://github.com/user-attachments/assets/46897389-7f93-4047-8dd5-fceca7f4b9ad)


Ui components in unity.

emilnorsker commented:
biog text fields and errors when not having a model node
![image](https://github.com/user-attachments/assets/46acf0c1-35c3-45cb-a4b7-15512bc5ce9c)

Documentation on hover for each field.
![image](https://github.com/user-attachments/assets/f34459a2-9b82-408c-add2-5272261727d3)

emilnorsker commented:
Due to the actor refactor, we can now use a callback to the unity side to trigger a signal. 


on the rust side it would look like this:
```rust

struct EmbeddingCallback {
    // eg. EmbeddingContext::OnEmbeddingRecieved(&self, embedding: Vec<f32>)
    callback: extern "C" fn(*mut c_void, *mut f32, i32),
    // eg. EmbeddingContext::new() as a pointer
    caller_ptr: *mut c_void,
}

impl chat::EmbeddingOutput for EmbeddingCallback {
    // We get a callback  from a static function that takes a object as its first argument.
    // This implicityly binds the callback to the object. 
    fn emit_embedding(&self, embedding: Vec<f32>) {
        let callback = unsafe { &self.callback };
        let caller_ptr = unsafe { self.caller_ptr };
        callback(caller_ptr, embedding.as_ptr() as *mut f32, embedding.len() as i32);
    }
}

/// This function creates a new embedding actor.
/// the embedding actor is responsible for generating embeddings for a given text.
#[no_mangle]
pub extern "C" fn create_embedding_worker(
    model_ptr: *mut c_void,
    caller_ptr: *mut c_void,
    callback: extern "C" fn(*mut c_void, *mut f32, i32),
    error_buf: *mut c_char,
) -> *mut c_void {
    if model_ptr.is_null() {
        copy_to_error_buf(error_buf, "Model pointer is null");
        return std::ptr::null_mut();
    }
    let model = unsafe { &mut *(model_ptr as *mut ModelObject) };

    let llm_params = llm::LLMActorParams {
        model: model.model.clone(),
        sampler_config: SamplerConfig::default(),
        stop_tokens: vec![],
        n_ctx: 4096,
        use_embeddings: true,
    };

    let (embed_tx, embed_rx) = tokio::sync::mpsc::channel(4096);
    // TODO: recieve a callback from the embedding worker to send the embedding back to the unity side
    let actor = chat::simple_embedding_loop(
        llm_params,
        embed_rx,
        Box::new(EmbeddingCallback { caller_ptr, callback }),
    );

    return Box::into_raw(Box::new(embed_tx)) as *mut c_void;
}

#[no_mangle]
pub extern "C" fn embed_text(
    embed_tx: *mut c_void,
    text: *const c_char,
    error_buf: *mut c_char,
 ) {
    let embed_tx = unsafe { &mut *(embed_tx as *mut tokio::sync::mpsc::Sender<String>) };

    let text_str: String = match unsafe { CStr::from_ptr(text).to_str() } {
        Ok(text_string) => text_string.to_owned(),
        Err(e) => {
            copy_to_error_buf(error_buf, &format!("Invalid UTF-8 in text: {}", e));
            return;
        }
    };

    match embed_tx.send(text_str) {
        Ok(_) => {
            println!("[DEBUG] embed_text - Text sent successfully");
        }
        Err(e) => {
            copy_to_error_buf(error_buf, &format!("Failed to send text: {}", e));
        }
    }
}
```
The C# side can look like this:

```csharp

public class Embedding : MonoBehaviour {
...
 [AOT.MonoPInvokeCallback(typeof(NativeBindings.EmbeddingCallback))]
        private static void OnEmbeddingCallback(IntPtr caller_ptr, IntPtr data, int length) {
            GCHandle handle = GCHandle.FromIntPtr(caller_ptr);
            Embedding instance = handle.Target as Embedding;
            
            float[] embedding = new float[length];
            Marshal.Copy(data, embedding, 0, length);
            
            instance._embeddingSignal?.SetResult(embedding);
            instance.onEmbeddingComplete.Invoke(embedding);
        }

        void Start() {
            _gcHandle = GCHandle.Alloc(this);
            var errorBuffer = new StringBuilder(2048);
            
            _workerContext = NativeBindings.create_embedding_worker(
                model.GetModel(),
                GCHandle.ToIntPtr(_gcHandle),
                OnEmbeddingCallback,
                errorBuffer
            );
            
            if (errorBuffer.Length > 0) {
                Debug.LogError(errorBuffer.ToString());
                enabled = false;
            }
        }

        public Awaitable<float[]> Embed(string text) {
            _embeddingSignal = new AwaitableCompletionSource<float[]>();
            
            var errorBuffer = new StringBuilder(2048);
            NativeBindings.embed_text(_workerContext, text, errorBuffer);
            
            if (errorBuffer.Length > 0) {
                Debug.LogError(errorBuffer.ToString());
            }
            
            return _embeddingSignal.Awaitable;
        }

** this is not the final implementation but rather an idea as to how it looks.
```

Note:  The callback struct that takes both a pointer and a callback function, as we can only pass static functions from C# to Rust through FFI, we need a user pointer to actually make this work. This approach opens a can of worms full of safety issues.

## Whether or not to use callbacks

Assumption: 
* The rest of the codebase is currently completely asynchronous (apart from the integration point with , thus we damage the codebase by not writing this as asynchronous as possible.
* 
 
pro/con:
+ We are not dependent on polling in the update loop at every frame.
+ Messages complete with immediacy, we don;t have to wait for the next frame.
+ More readable flow.
- Static routing requires `GCHandles` which adds complex memory handling on the C# side.
- Callbacks may execute on background threads, requiring thread synchronization.
- Tests are slightly harder to implement on the rust side as the callback need to be  

This comes as a contrast to using a message queue system which has its own advantages and disadvantages:

+ Better decoupling between C# and Rust components.
- Requires polling mechanism to check for new messages.
- May complicate debugging with less direct execution flow.
+ Simpler memory model, we don't have to explicitly  take c are of memory on the C# side.

AsbjornOlling commented on nobodywho/core/src/chat.rs:
```diff
@@ -36,7 +39,11 @@ pub enum ChatLoopError {
 }
 
-pub trait ChatOutput {
+// This trait requires the consumers to be both send and sync.
```
Fixed. Neither Send nor Sync are required.

AsbjornOlling commented on nobodywho/core/default.nix:
Done.

AsbjornOlling commented on nobodywho/unity/default.nix:
Fixed.

AsbjornOlling commented on nobodywho/unity/src/Runtime/NobodyWhoChat.cs:
```diff
@@ -0,0 +1,92 @@
+        public string grammar;
+
+        [Header("Events")]
+        public UnityEvent<string> onToken = new UnityEvent<string>();
```
It's called `response_updated` and `response_finished` in godot, but it's called `onToken` and `onComplete` in Unity?

Maybe we should rename things, so they're called the same thing in all of our plugins?
That might actually make the documentation be somewhat usable across engines.

If there's a good reason to change the name of the godot signals, I'm good with doing that. Otherwise, I think we should stick with the existing names, to avoid having to release a breaking change.

AsbjornOlling commented on nobodywho/core/Cargo.toml:
```diff
@@ -4,6 +4,7 @@ version = "2.0.0"
 edition = "2021"
 
 [lib]
+crate-type = ["cdylib", "rlib"]
```
Why is this necessary? I don't think we're building the core crate by itself.

AsbjornOlling commented on nobodywho/core/src/chat.rs:
```diff
@@ -11,19 +12,22 @@ use llama_cpp_2::model::LlamaModel;
 
 pub struct ChatHandle {
     msg_tx: std::sync::mpsc::Sender<ChatMsg>,
+    pub ready: Arc<AtomicBool>,
```
This flag isn't needed after rewriting the FFI using interoptopus. We should get rid of the ready atomicbool.

AsbjornOlling commented on nobodywho/core/src/chat.rs:
```diff
@@ -73,18 +77,27 @@ fn run_worker(
+    //      if we don't do this, we get memory errors in the unity integration
+    //      inside the ctx.decode call in `write_until_done`
+    //      very whack. feel free to fix. hours wasted on this: many
+    ready.store(true, Ordering::SeqCst);
```
Already mentioned, but we should get rid of this.

AsbjornOlling commented on nobodywho/core/src/lib.rs:
```diff
@@ -6,6 +6,10 @@ pub mod llm;
 // pub mod llmactor;
 pub mod sampler_config;
 
+pub fn send_llamacpp_logs_to_tracing() {
```
Might need to rebase on main. This line is already on the lastest main.

emilnorsker commented on nobodywho/godot/integration-test/nobodywho.gdextension:
yeah i had edited the wrong file .gdextension file. Good catch :)

emilnorsker commented on nobodywho/unity/src/Runtime/NobodyWhoChat.cs:
```diff
@@ -0,0 +1,262 @@
+        public string systemPrompt = "";
+
+        [Header("Configuration")]
+        public string stopWords = "";
```
```
            EditorGUILayout.PropertyField(
                serializedObject.FindProperty("stopWords"),
                new GUIContent(
                    "Stop Words",
                    "Stop words to stop generation at these specified tokens, seperated by commas"
                )
            );
```

there is a tool tip. for it.

emilnorsker commented on nobodywho/unity/README.md:
```diff
@@ -0,0 +1,42 @@
+
+Unity's licensing is a bit of a pain, so we cant create a properly sandboxed dev environment.
+Due to this, we use a Justfile to run the Unity editor with a custom build script.
+The unity executable is downloaded by the `nix develop` command. But the license needs to be downloaded manually.
```
> I think this is misinformed. Nix develop does not download or install a unity editor.
The unity/default.nix file just expects there to already be a unity executable with a very specific version in your home folder.

correct, i forgot to update this when i changed the nix setup- Unity is not downloaded (the unity license file will generate a machine specific id, that will change based on whether or not we are in a nix shell) - thus rendering the  downloading a unity executable useless. 

i think that we could try to patch that in a future edition as with how Game-ci's actions do it. the license will still have to be downloaded.
